pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,275 @@
+			log.Debug("msg", "syncing new stats", "error", err.Error())
+		}
+	}
+}
+
+// metricsVisitor fills the provided newStats with most recent metric values.
+func (t *engine) metricsVisitor(newStats *telemetryStats) error {
+	t.metricMux.RLock()
+	defer t.metricMux.RUnlock()
+	if len(t.metrics) == 0 {
+		return nil
+	}
+	for _, stat := range t.metrics {
+		underlyingValue, err := extractMetricValue(stat.internal())
+		if err != nil {
+			return fmt.Errorf("extracting metric value of stat '%s': %w", stat.name(), err)
+		}
+		switch stat.name() {
+		case "telemetry_metrics_ingested_samples":
+			newStats.samplesIngested = underlyingValue
+		case "telemetry_metrics_queries_failed":
+			newStats.promqlQueriesFailed = underlyingValue
+		case "telemetry_metrics_queries_executed":
+			newStats.promqlQueriesExecuted = underlyingValue
+		case "telemetry_metrics_queries_timed_out":
+			newStats.promqlQueriesTimedout = underlyingValue
+		case "telemetry_traces_queries_executed":
+			newStats.traceQueriesExecuted = underlyingValue
+		case "telemetry_traces_dependency_queries_executed":
+			newStats.traceDependencyQueriesExecuted = underlyingValue
+		default:
+			log.Error("msg", fmt.Sprintf("unregistered stat name: %s", stat.name()))
+		}
+	}
+	return nil
+}
+
+func extractMetricValue(metric prometheus.Metric) (float64, error) {
+	var internal io_prometheus_client.Metric
+	if err := metric.Write(&internal); err != nil {
+		return 0, fmt.Errorf("error writing metric: %w", err)
+	}
+	if internal.Gauge != nil {
+		return internal.Gauge.GetValue(), nil
+	} else if internal.Counter != nil {
+		return internal.Counter.GetValue(), nil
+	}
+	return 0, fmt.Errorf("both Gauge and Counter are nil")
+}
+
+// RegisterMetric registers a counter or gauge metric for telemetry purpose.
+func (t *engine) RegisterMetric(columnName string, metric prometheus.Metric) error {
+	ok, err := isCounterOrGauge(metric)
+	if err != nil {
+		return fmt.Errorf("is counter or gauge: %w", err)
+	}
+	if !ok {
+		return fmt.Errorf("metric should only be of type Counter or Gauge")
+	}
+	t.metricMux.Lock()
+	defer t.metricMux.Unlock()
+	t.metrics = append(t.metrics, telemetryMetric{stat: columnName, metric: metric})
+	return nil
+}
+
+func isCounterOrGauge(metric prometheus.Metric) (bool, error) {
+	var temp io_prometheus_client.Metric
+	if err := metric.Write(&temp); err != nil {
+		return false, fmt.Errorf("writing metric: %w", err)
+	}
+	if temp.Gauge != nil {
+		return true, nil
+	} else if temp.Counter != nil {
+		return true, nil
+	}
+	return false, nil
+}
+
+// syncInfoTable stats with promscale_instance_information table.
+func syncInfoTable(conn pgxconn.PgxConn, uuid [16]byte, s telemetryStats) error {
+	lastUpdated := time.Now()
+
+	pgUUID := new(pgtype.UUID)
+	if err := pgUUID.Set(uuid); err != nil {
+		return fmt.Errorf("setting pg-uuid: %w", err)
+	}
+
+	query := `INSERT INTO _ps_catalog.promscale_instance_information
```

Comment:
The schema name is hardcoded here, but isn't it technically configurable?

Reply:
No, this is a static schema which is not configurable upon user. This schema will be further expanded in future, by moving contents of `_prom_catalog` here.

Reply:
Still, lets use the `schema.PSCatalog` here instead of hardcoding. While unlikely we will change the name, its safer and easier to navigate code when looking for code that uses that schema.